README.md
can_prius_ros
================================

A ROS wrapper around libcan-prius


changelog
can-prius-ros (0.1-33) precise; urgency=low

  * cleaning the code 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 26 Sep 2014 10:26:59 +0200

can-prius-ros (0.1-32) precise; urgency=low

  * porting to trusty 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 25 Sep 2014 16:13:56 +0200

can-prius-ros (0.1-31) precise; urgency=low

  * issues with groovy 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 27 Mar 2014 13:37:41 +0100

can-prius-ros (0.1-30) precise; urgency=low

  * should now work with groovy 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 26 Mar 2014 16:31:23 +0100

can-prius-ros (0.1-29) precise; urgency=low

  * another test with groovy

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 26 Mar 2014 11:22:31 +0100

can-prius-ros (0.1-28) precise; urgency=low

  * adding pkg-config 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 25 Mar 2014 17:00:35 +0100

can-prius-ros (0.1-27) precise; urgency=low

  * update version

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Sun, 14 Jul 2013 15:35:40 +0000

can-prius-ros (0.1-26) precise; urgency=low

  * groovy distribution 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 06 Jun 2013 08:54:14 +0000

can-prius-ros (0.1-25) precise; urgency=low

  * changing frame_id 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 05 Jun 2013 17:35:55 +0000

can-prius-ros (0.1-24) precise; urgency=low

  * make the node respawn 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 30 May 2013 17:59:16 +0000

can-prius-ros (0.1-23) precise; urgency=low

  * uipdate of remae for mixing up ROS distributions 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 30 May 2013 13:49:37 +0000

can-prius-ros (0.1-22) precise; urgency=low

  * removing nullptr 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 29 May 2013 14:39:29 +0000

can-prius-ros (0.1-21) precise; urgency=low

  * missing dependencies 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 29 May 2013 07:57:26 +0000

can-prius-ros (0.1-20) precise; urgency=low

  * distribution on PPA 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 28 May 2013 18:46:36 +0000

can-prius-ros (0.1-19) precise; urgency=low

  * changing /dev/cpc_usb_0 to /dev/cpc_usb0 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 28 May 2013 17:08:47 +0000

can-prius-ros (0.1-18) precise; urgency=low

  * changing library name 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 28 May 2013 09:16:55 +0000

can-prius-ros (0.1-17) precise; urgency=low

  * frequency tuning, packet counters, diagnose improved

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Sat, 25 May 2013 10:44:34 +0000

can-prius-ros (0.1-16) precise; urgency=low

  * changing hardware ID for diagnostics 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 24 May 2013 10:23:31 +0000

can-prius-ros (0.1-15) precise; urgency=low

  * changing parameters 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

can-prius-ros (0.1-14) precise; urgency=low

  * changing names 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 19:52:42 +0000

can-prius-ros (0.1-13) precise; urgency=low

  * porting to remake

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 17:43:40 +0000

can-prius-ros (0.1-12) lucid; urgency=low

  * adding diagnostics

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-11) lucid; urgency=low

  * adding diagnostics

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-10) lucid; urgency=low

  * adding dependency

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-9) lucid; urgency=low

  * adding yaml file

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-8) lucid; urgency=low

  * changing ros::ok() to nh.ok()

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-7) lucid; urgency=low

  * minor change

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-6) lucid; urgency=low

  * adding steering message

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-5) lucid; urgency=low

  * indentation

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-4) lucid; urgency=low

  * changed license to LGPL

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-3) lucid; urgency=low

  * refactoring

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-2) lucid; urgency=low

  * adding the first 2 messages

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-1) lucid; urgency=low

  * sketching up the ROS wrapper

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200

can-prius-ros (0.1-0) lucid; urgency=low

  * Initial commit

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 13 May 2013 14:36:23 +0200


src/can_prius_node.cpp
/******************************************************************************
 * Copyright (C) 2013 by Jerome Maye                                          *
 * jerome.maye@gmail.com                                                      *
 *                                                                            *
 * This program is free software; you can redistribute it and/or modify       *
 * it under the terms of the Lesser GNU General Public License as published by*
 * the Free Software Foundation; either version 3 of the License, or          *
 * (at your option) any later version.                                        *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 * Lesser GNU General Public License for more details.                        *
 *                                                                            *
 * You should have received a copy of the Lesser GNU General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 ******************************************************************************/

/** \file can_prius_node.cpp
    \brief This file is the ROS node for CAN Prius.
  */

#include <ros/ros.h>

#include "janeth/CANPriusNode.h"

int main(int argc, char** argv) {
  ros::init(argc, argv, "can_prius");
  ros::NodeHandle nh("~");
  try {
    janeth::CANPriusNode cn(nh);
    cn.spin();
  }
  catch (const std::exception& e) {
    ROS_ERROR_STREAM("Exception: " << e.what());
    return 1;
  }
  catch (...) {
    ROS_ERROR_STREAM("Unknown Exception");
    return 1;
  }
  return 0;
}


src/can_prius/lib/CANPriusNode.h
/******************************************************************************
 * Copyright (C) 2013 by Jerome Maye                                          *
 * jerome.maye@gmail.com                                                      *
 *                                                                            *
 * This program is free software; you can redistribute it and/or modify       *
 * it under the terms of the Lesser GNU General Public License as published by*
 * the Free Software Foundation; either version 3 of the License, or          *
 * (at your option) any later version.                                        *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 * Lesser GNU General Public License for more details.                        *
 *                                                                            *
 * You should have received a copy of the Lesser GNU General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 ******************************************************************************/

/** \file CANPriusNode.h
    \brief This file defines the CANPriusNode class which implements
           the CAN Prius node.
  */

#ifndef CAN_PRIUS_NODE_H
#define CAN_PRIUS_NODE_H

#include <string>
#include <memory>

#include <ros/ros.h>
#include <diagnostic_updater/diagnostic_updater.h>

class FrontWheelsSpeed;
class RearWheelsSpeed;
class Steering1;
class CANConnection;

namespace diagnostic_updater {
  class HeaderlessTopicDiagnostic;
}

namespace prius {

  /** The class CANPriusNode implements the CAN Prius node.
      \brief CAN Prius node
    */
  class CANPriusNode {
  public:
    /** \name Constructors/destructor
      @{
      */
    /// Constructor
    CANPriusNode(const ros::NodeHandle& nh);
    /// Copy constructor
    CANPriusNode(const CANPriusNode& other) = delete;
    /// Copy assignment operator
    CANPriusNode& operator = (const CANPriusNode& other) = delete;
    /// Move constructor
    CANPriusNode(CANPriusNode&& other) = delete;
    /// Move assignment operator
    CANPriusNode& operator = (CANPriusNode&& other) = delete;
    /// Destructor
    virtual ~CANPriusNode();
    /** @}
      */

    /** \name Methods
      @{
      */
    /// Spin once
    void spin();
    /** @}
      */

  protected:
    /** \name Protected methods
      @{
      */
    /// Publishes the front wheel speed
    void publishFrontWheelsSpeed(const ros::Time& timestamp,
      const FrontWheelsSpeed& fws);
    /// Publishes the rear wheel speed
    void publishRearWheelsSpeed(const ros::Time& timestamp,
      const RearWheelsSpeed& rws);
    /// Publishes the steering1
    void publishSteering1(const ros::Time& timestamp, const Steering1& st);
    /// Diagnose the CAN connection
    void diagnoseCANConnection(diagnostic_updater::DiagnosticStatusWrapper&
      status);
    /// Retrieves parameters
    void getParameters();
    /** @}
      */

    /** \name Protected members
      @{
      */
    /// ROS node handle
    ros::NodeHandle _nodeHandle;
    /// Front wheel speed publisher
    ros::Publisher _frontWheelsSpeedPublisher;
    /// Rear wheel speed publisher
    ros::Publisher _rearWheelsSpeedPublisher;
    /// Steering1 publisher
    ros::Publisher _steering1Publisher;
    /// Frame ID
    std::string _frameId;
    /// CAN device
    std::string _canDeviceStr;
    /// CAN connection
    std::shared_ptr<CANConnection> _canConnection;
    /// Retry timeout for CAN
    double _retryTimeout;
    /// Diagnostic updater
    diagnostic_updater::Updater _updater;
    /// Frequency diagnostic for front wheels speed
    std::shared_ptr<diagnostic_updater::HeaderlessTopicDiagnostic> _fwsFreq;
    /// Front wheels speed minimum frequency
    double _fwsMinFreq;
    /// Front wheels speed maximum frequency
    double _fwsMaxFreq;
    /// Frequency diagnostic for rear wheels speed
    std::shared_ptr<diagnostic_updater::HeaderlessTopicDiagnostic> _rwsFreq;
    /// Rear wheels speed minimum frequency
    double _rwsMinFreq;
    /// Rear wheels speed maximum frequency
    double _rwsMaxFreq;
    /// Frequency diagnostic for steering1
    std::shared_ptr<diagnostic_updater::HeaderlessTopicDiagnostic> _st1Freq;
    /// Steering1 1minimum frequency
    double _st1MinFreq;
    /// Steering1 maximum frequency
    double _st1MaxFreq;
    /// Queue depth
    int _queueDepth;
    /// Front wheels speed packet counter
    long _fwsPacketCounter;
    /// Rear wheels speed packet counter
    long _rwsPacketCounter;
    /// Steering1 packet counter
    long _st1PacketCounter;
    /** @}
      */

  };

}

#endif // CAN_PRIUS_NODE_H


src/can_prius/lib/CANPriusNode.cpp
/******************************************************************************
 * Copyright (C) 2013 by Jerome Maye                                          *
 * jerome.maye@gmail.com                                                      *
 *                                                                            *
 * This program is free software; you can redistribute it and/or modify       *
 * it under the terms of the Lesser GNU General Public License as published by*
 * the Free Software Foundation; either version 3 of the License, or          *
 * (at your option) any later version.                                        *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 * Lesser GNU General Public License for more details.                        *
 *                                                                            *
 * You should have received a copy of the Lesser GNU General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 ******************************************************************************/

#include "CANPriusNode.h"

#include <diagnostic_updater/publisher.h>

#include <boost/shared_ptr.hpp>
#include <boost/make_shared.hpp>

#include <libcan-prius/com/CANConnection.h>
#include <libcan-prius/sensor/PRIUSReader.h>
#include <libcan-prius/types/PRIUSMessage.h>
#include <libcan-prius/types/FrontWheelsSpeed.h>
#include <libcan-prius/types/RearWheelsSpeed.h>
#include <libcan-prius/types/Speed1.h>
#include <libcan-prius/types/Speed2.h>
#include <libcan-prius/types/Speed3.h>
#include <libcan-prius/types/Steering1.h>
#include <libcan-prius/types/Steering2.h>
#include <libcan-prius/types/Brakes.h>
#include <libcan-prius/types/Acceleration1.h>
#include <libcan-prius/types/Acceleration2.h>
#include <libcan-prius/exceptions/IOException.h>
#include <libcan-prius/base/Timer.h>

#include "can_prius/FrontWheelsSpeedMsg.h"
#include "can_prius/RearWheelsSpeedMsg.h"
#include "can_prius/Steering1Msg.h"

namespace prius {

/******************************************************************************/
/* Constructors and Destructor                                                */
/******************************************************************************/

  CANPriusNode::CANPriusNode(const ros::NodeHandle& nh) :
      _nodeHandle(nh),
      _fwsPacketCounter(0),
      _rwsPacketCounter(0),
      _st1PacketCounter(0) {
    getParameters();
    _frontWheelsSpeedPublisher =
      _nodeHandle.advertise<can_prius::FrontWheelsSpeedMsg>(
      "front_wheels_speed", _queueDepth);
    _rearWheelsSpeedPublisher =
      _nodeHandle.advertise<can_prius::RearWheelsSpeedMsg>(
      "rear_wheels_speed", _queueDepth);
    _steering1Publisher =
      _nodeHandle.advertise<can_prius::Steering1Msg>(
      "steering1", _queueDepth);
    _updater.setHardwareID("Toyota PRIUS CAN bus");
    _updater.add("CAN connection", this, &CANPriusNode::diagnoseCANConnection);
    _fwsFreq = std::make_shared<diagnostic_updater::HeaderlessTopicDiagnostic>(
      "front_wheels_speed", _updater,
      diagnostic_updater::FrequencyStatusParam(&_fwsMinFreq, &_fwsMaxFreq,
      0.1, 10));
    _rwsFreq = std::make_shared<diagnostic_updater::HeaderlessTopicDiagnostic>(
      "rear_wheels_speed", _updater,
      diagnostic_updater::FrequencyStatusParam(&_rwsMinFreq, &_rwsMaxFreq,
      0.1, 10));
    _st1Freq = std::make_shared<diagnostic_updater::HeaderlessTopicDiagnostic>(
      "steering1", _updater,
      diagnostic_updater::FrequencyStatusParam(&_st1MinFreq, &_st1MaxFreq,
      0.1, 10));
    _updater.force_update();
  }

  CANPriusNode::~CANPriusNode() {
  }

/******************************************************************************/
/* Methods                                                                    */
/******************************************************************************/

  void CANPriusNode::publishFrontWheelsSpeed(const ros::Time& timestamp,
      const FrontWheelsSpeed& fws) {
    if (_frontWheelsSpeedPublisher.getNumSubscribers() > 0) {
      auto fwsMsg = boost::make_shared<can_prius::FrontWheelsSpeedMsg>();
      fwsMsg->header.stamp = timestamp;
      fwsMsg->header.frame_id = _frameId;
      fwsMsg->header.seq = _fwsPacketCounter++;
      fwsMsg->Right = fws.mRight;
      fwsMsg->Left = fws.mLeft;
      _frontWheelsSpeedPublisher.publish(fwsMsg);
    }
    _fwsFreq->tick();
  }

  void CANPriusNode::publishRearWheelsSpeed(const ros::Time& timestamp,
      const RearWheelsSpeed& rws) {
    if (_rearWheelsSpeedPublisher.getNumSubscribers() > 0) {
      auto rwsMsg = boost::make_shared<can_prius::RearWheelsSpeedMsg>();
      rwsMsg->header.stamp = timestamp;
      rwsMsg->header.frame_id = _frameId;
      rwsMsg->header.seq = _rwsPacketCounter++;
      rwsMsg->Right = rws.mRight;
      rwsMsg->Left = rws.mLeft;
      _rearWheelsSpeedPublisher.publish(rwsMsg);
    }
    _rwsFreq->tick();
  }

  void CANPriusNode::publishSteering1(const ros::Time& timestamp,
      const Steering1& st) {
    if (_steering1Publisher.getNumSubscribers() > 0) {
      auto stMsg = boost::make_shared<can_prius::Steering1Msg>();
      stMsg->header.stamp = timestamp;
      stMsg->header.frame_id = _frameId;
      stMsg->header.seq = _st1PacketCounter++;
      stMsg->value = st.mValue;
      _steering1Publisher.publish(stMsg);
    }
    _st1Freq->tick();
  }

  void CANPriusNode::diagnoseCANConnection(
      diagnostic_updater::DiagnosticStatusWrapper& status) {
    if (_canConnection && _canConnection->isOpen())
      status.summaryf(diagnostic_msgs::DiagnosticStatus::OK,
        "CAN connection opened on %s.",
        _canConnection->getDevicePathStr().c_str());
    else
     status.summaryf(diagnostic_msgs::DiagnosticStatus::ERROR,
      "CAN connection closed on %s.", _canDeviceStr.c_str());
  }

  void CANPriusNode::spin() {
    _canConnection = std::make_shared<CANConnection>(_canDeviceStr);
    PRIUSReader reader(*_canConnection);
    Timer timer;
    while (_nodeHandle.ok()) {
      try {
        std::shared_ptr<PRIUSMessage> message = reader.readMessage();
        const ros::Time timestamp = ros::Time::now();
        if (message->instanceOf<FrontWheelsSpeed>()) {
          const FrontWheelsSpeed& fws = message->typeCast<FrontWheelsSpeed>();
          publishFrontWheelsSpeed(timestamp, fws);
        }
        else if (message->instanceOf<RearWheelsSpeed>()) {
          const RearWheelsSpeed& rws = message->typeCast<RearWheelsSpeed>();
          publishRearWheelsSpeed(timestamp, rws);
        }
        else if (message->instanceOf<Speed1>()) {
          const Speed1& sp = message->typeCast<Speed1>();
        }
        else if (message->instanceOf<Speed2>()) {
          const Speed2& sp = message->typeCast<Speed2>();
        }
        else if (message->instanceOf<Speed3>()) {
          const Speed3& sp = message->typeCast<Speed3>();
        }
        else if (message->instanceOf<Steering1>()) {
          const Steering1& st = message->typeCast<Steering1>();
          publishSteering1(timestamp, st);
        }
        else if (message->instanceOf<Steering2>()) {
          const Steering2& st = message->typeCast<Steering2>();
        }
        else if (message->instanceOf<Brakes>()) {
          const Brakes& b = message->typeCast<Brakes>();
        }
        else if (message->instanceOf<Acceleration1>()) {
          const Acceleration1& a = message->typeCast<Acceleration1>();
        }
        else if (message->instanceOf<Acceleration2>()) {
          const Acceleration2& a = message->typeCast<Acceleration2>();
        }
      } 
      catch (const IOException& e) {
        ROS_WARN_STREAM("IOException: " << e.what());
        ROS_WARN_STREAM("Retrying in " << _retryTimeout << " [s]");
        timer.sleep(_retryTimeout);
      }
      _updater.update();
      ros::spinOnce();
    }
  }

  void CANPriusNode::getParameters() {
    _nodeHandle.param<std::string>("ros/frame_id", _frameId,
      "/can_link");
    _nodeHandle.param<int>("ros/queue_depth", _queueDepth, 100);
    _nodeHandle.param<std::string>("connection/can_device", _canDeviceStr,
      "/dev/cpc_usb0");
    _nodeHandle.param<double>("connection/retry_timeout", _retryTimeout, 1);
    _nodeHandle.param<double>("diagnostics/fws_min_freq", _fwsMinFreq, 48);
    _nodeHandle.param<double>("diagnostics/fws_max_freq", _fwsMaxFreq, 72);
    _nodeHandle.param<double>("diagnostics/rws_min_freq", _rwsMinFreq, 48);
    _nodeHandle.param<double>("diagnostics/rws_max_freq", _rwsMaxFreq, 72);
    _nodeHandle.param<double>("diagnostics/st1_min_freq", _st1MinFreq, 24);
    _nodeHandle.param<double>("diagnostics/st1_max_freq", _st1MaxFreq, 36);
  }

}
